axlearn/common/attention.py

```diff
@@ -2355,6 +2356,12 @@ class Config(BaseLayer.Config):
         # Ref: https://github.com/google/praxis/blob/main/praxis/layers/transformers.py#L1129
         # TODO (bwzhang@) Adding a unittest for the hybridnorm.
         structure: str = "prenorm"
+        # If not None, how to partition pre norm activation values.
+        prenorm_partition_spec: Optional[tuple[Optional[str]]] = None
+        # If not None, how to partition pre attention activation values.
+        preattention_partition_spec: Optional[tuple[Optional[str]]] = None
+        # If not None, how to partition post attention activation values.
+        postattention_partition_spec: Optional[tuple[Optional[str]]] = None
```

Comment:
How is this different from setting https://github.com/apple/axlearn/blob/2134a258a869f679e4dd2625565a2ca693126495/axlearn/common/attention.py#L3319?

Reply:
+1, I also have a high level question, should the sharding config auto-derived and optimized by XLA for most of the intermediary activations, based on weights and outputs etc? 
I'd like to see a concrete example to understand why explicit sharding is needed here.

Reply:
OpenXLA ShardingPropagation often makes non optimal decisions in propagation. SPMD partitioner then partitions based on what HloSharding was propagated to each HloInstruction, so if your sharding propagation is not optimal then you will see collectives generated such as all-to-all to resolve the sharding conflicts.

An example is if you set the partition spec for premlp or preattention to be ((fsdp, data), model, None) to force the computation to be done in sequence parallel, then ShardingPropagation pass will propagate that HloSharding spec down to the matmul causing an all-to-all collective. 

This is not as optimal as an all-gather of the sequence dimension before the matmul.